src/Emulators/M68KEmulator.cc

```diff
@@ -3383,20 +3383,182 @@ const M68KEmulator::OpcodeImplementation M68KEmulator::fns[0x10] = {
 
 ////////////////////////////////////////////////////////////////////////////////
 
+
+static bool is_valid_macsbug_symbol_char(char ch) {
+  // "Building and Managing Programs in MPW", chapter B-25:
+  //
+  //    A valid MacsBug symbol consists of _ characters, % characters, spaces,
+  //    digits, and uppercase and lowercase letters
+  //
+  // "Macsbug Reference and Debugging Guide", page 367:
+  //
+  //    Valid characters for procedure names are a–z, A–Z, 0–9, underscore (_), 
+  //    percent (%), period (.), and space
+  
+  // Do not use 'isalpha' etc. as they take the current locale into account
+  return  ch == '_' ||
+          ch == '%' ||
+          ch == '.' ||
+          ch == ' ' ||
+          (ch >= '0' && ch <= '9') ||
+          (ch >= 'A' && ch <= 'Z') ||
+          (ch >= 'a' && ch <= 'z');
+}
+
+
+static bool try_decode_macsbug_symbol_part(StringReader& r, string& symbol, uint16_t symbol_length) {
+  if (r.remaining() < symbol_length) {
+    return false;
+  }
+  
+  for (uint16_t i = 0; i < symbol_length; ++i) {
+    uint8_t ch = r.get_u8();
+    if (!is_valid_macsbug_symbol_char(ch)) {
+      return false;
+    }
+    symbol += ch;
+  }
+  
+  return true;
+}
+
+
+static bool try_decode_macsbug_symbol(StringReader& r, string& symbol, uint16_t& num_constants) {

```

Comment:
I think it would be cleaner to have this function return a struct containing symbol and num_constants (and perhaps a vector of the constants later). It could throw out_of_range if decoding fails, or just return a struct with empty values in that case.

Reply:
Originally I had planned to return an empty string, but (incorrectly) assumed that the encoding allowed for empty names, which is why I opted to return a bool.

I don't think an exception is the right choice here, as `try_decode_macsbug_symbol` is called on every instruction following a function return and is allowed to fail. So the exception wouldn't be used for error handling, but for control flow, which is generally considered bad.

I pushed a change to make `try_decode_macsbug_symbol` return a struct.